src/main/java/com/researchspace/api/v1/controller/GalleryFilestoresApiController.java

```diff
@@ -0,0 +1,217 @@
+  public List<NfsFileStoreInfo> getUserFilestores(@RequestAttribute(name = "user") User user) {
+    assertFilestoresApiEnabled(user);
+    return nfsManager.getFileStoreInfosForUser(user);
+  }
+
+  @Override
+  public ApiNfsRemotePathBrowseResult browseFilestore(
+      @PathVariable Long filestoreId,
+      @RequestParam(value = "remotePath", required = false) String browsePath,
+      @RequestAttribute(name = "user") User user)
+      throws IOException {
+
+    assertFilestoresApiEnabled(user);
+    NfsFileStore filestore = nfsManager.getNfsFileStore(filestoreId);
+    NfsFileSystem filesystem = filestore.getFileSystem();
+    NfsClient nfsClient = credentialsStore.getNfsClientWithStoredCredentials(user, filesystem);
+
+    String combinedPath = filestore.getPath();
+    if (StringUtils.isNotEmpty(browsePath)) {
+      combinedPath += browsePath;
+    }
+
+    NfsFileTreeNode fileTree = nfsClient.createFileTree(combinedPath, null, filestore);
+    return getRemotePathBrowseResult(filesystem, nfsClient, fileTree);
+  }
+
+  private ApiNfsRemotePathBrowseResult getRemotePathBrowseResult(
+      NfsFileSystem filesystem, NfsClient nfsClient, NfsFileTreeNode fileTree) {
+
+    return new ApiNfsRemotePathBrowseResult(
+        fileTree,
+        filesystem.toFileSystemInfo(),
+        nfsClient.supportsExtraDirs() && !filesystem.fileSystemRequiresUserRootDirs(),
+        nfsClient.supportsCurrentDir(),
+        nfsClient.getUsername());
+  }
+
+  @Override
+  public void downloadFromFilestore(
+      @PathVariable Long filestoreId,
+      @RequestParam(value = "remotePath", required = false) String remotePath,
+      @RequestParam(name = "remoteId", required = false) Long remoteId,
+      @RequestAttribute(name = "user") User user,
+      HttpServletResponse response)
+      throws IOException {
+
+    assertFilestoresApiEnabled(user);
+    if (StringUtils.isEmpty(remotePath) && remoteId == null) {
+      throw new IllegalArgumentException("Neither 'remotePath' nor 'remoteId' param is provided");
+    }
+
+    NfsFileStore filestore = nfsManager.getNfsFileStore(filestoreId);
+    NfsFileSystem filesystem = filestore.getFileSystem();
+    NfsClient nfsClient = credentialsStore.getNfsClientWithStoredCredentials(user, filesystem);
+
+    String fullPath = filestore.getAbsolutePath(remotePath);
+    NfsFileDetails nfsFileDetails =
+        nfsFileHandler.downloadNfsFileToRSpace(new NfsTarget(fullPath, remoteId), nfsClient);
+    File downloadedFile = nfsFileDetails.getLocalFile();
+    log.info("downloaded to: " + downloadedFile.getCanonicalPath());
+
+    response.setContentType("application/octet-stream");
+    response.setContentLength((int) downloadedFile.length());
+    response.setHeader(
+        "Content-Disposition", "attachment; filename=\"" + downloadedFile.getName() + "\"");
+    try (InputStream is = new FileInputStream(downloadedFile);
+        ServletOutputStream out = response.getOutputStream()) {
+      IOUtils.copy(is, out);
```

Comment:
Is the exception management (i.e.: disk full, file name already existing, or any other issue that can happen in a filesystem) taken care somewhere else?

Reply:
The download code is actually part of pre-existing `NfsFileHandler` that was used in old UI, here the API code just assumes the file is there and streams it to servlet response, that should always be fine. 

Looking into implementation in `NfsFileHandler.downloadNfsFileToRSpace()` - it downloads file into a new temp folder each time, so that should cover some potential problems like name conflicts. It doesn't check the remaining free disk space though.